### PR TITLE
feat: ability to show UI Bar values as percentage

### LINKED ISSUE #1431: feature: setting to toggle UI's EXP bar value to % (With Decimals)
### Description

A way to change the exp string for % at the client config, with a way to put how much zeros needed (decimals).

Example:
Old method: current EXP / EXP to next level  ( 1 / 100000 )
New method: % exp ( 0,0000%)

### Intended Use-Case

Many players like me, dont like the astronomic numbers after lv up, with % at current exp is less stressing,

### Duplicate Check

- [X] This feature request is not a duplicate to the best of my knowledge.

### PR DESCRIPTION
**Pull request by Arufonsu**
- chore: some things were reorganized around the modified code.
- new feat: ability to show UI Bar values as percentage.
- If enabled, will display the toggled bar's as percentage and it's tooltip will show the actual value instead.
- If disabled, will display the toggled bar's actual value and it's tooltip will show the percentage instead.
- should resolve #1431 

**Assets PR :** https://github.com/AscensionGameDev/Intersect-Assets/pull/22

### Preview

https://user-images.githubusercontent.com/17498701/189506575-57de6696-52cf-4656-be1c-4428e3d7e366.mp4

### CODE REVIEW

File: Intersect.Client.Framework/Database/GameDatabase.cs [hunk 1]
```diff
@@ -26,17 +23,25 @@ public abstract partial class GameDatabase
...
+        public bool MyOverheadInfo { get; set; }
 
-        public bool MyOverheadInfo;
+        public bool NpcOverheadInfo { get; set; }
+
+        public bool PartyMemberOverheadInfo { get; set; }
+
+        public bool PlayerOverheadInfo { get; set; }
+
+        public bool HideOthersOnWindowOpen { get; set; }
```
Comment (lodicolo) [inline Intersect.Client.Framework/Database/GameDatabase.cs:38]: Why was this moved? It made reading this diff harder than necessary.

File: Intersect.Client.Framework/Database/GameDatabase.cs [hunk 2]
```diff
@@ -26,17 +23,25 @@ public abstract partial class GameDatabase
...
 
-        public bool NpcOverheadInfo;
+        public bool UiExpToPercentage { get; set; }
 
-        public bool PartyMemberOverheadInfo;
+        public bool UiHpToPercentage { get; set; }
```
Comment (lodicolo) [inline Intersect.Client.Framework/Database/GameDatabase.cs:42]: `ShowHealthAsPercentage`

File: Intersect.Client.Framework/Database/GameDatabase.cs [hunk 3]
```diff
@@ -26,17 +23,25 @@ public abstract partial class GameDatabase
...
+        public bool NpcOverheadInfo { get; set; }
+
+        public bool PartyMemberOverheadInfo { get; set; }
+
+        public bool PlayerOverheadInfo { get; set; }
+
+        public bool HideOthersOnWindowOpen { get; set; }
 
-        public bool NpcOverheadInfo;
+        public bool UiExpToPercentage { get; set; }
```
Comment (lodicolo) [inline Intersect.Client.Framework/Database/GameDatabase.cs:40]: `ShowExperienceAsPercentage`

File: Intersect.Client.Framework/Database/GameDatabase.cs [hunk 4]
```diff
@@ -26,17 +23,25 @@ public abstract partial class GameDatabase
...
 
-        public bool PartyMemberOverheadInfo;
+        public bool UiHpToPercentage { get; set; }
 
-        public bool PlayerOverheadInfo;
+        public bool UiMpToPercentage { get; set; }
```
Comment (lodicolo) [inline Intersect.Client.Framework/Database/GameDatabase.cs:44]: `ShowManaAsPercentage`

File: Intersect.Client.Framework/Database/GameDatabase.cs [hunk 5]
```diff
@@ -76,6 +81,9 @@ public virtual void LoadPreferences()
             NpcOverheadInfo = LoadPreference("NpcOverheadInfo", true);
             PartyMemberOverheadInfo = LoadPreference("PartyMemberOverheadInfo", true);
             PlayerOverheadInfo = LoadPreference("PlayerOverheadInfo", true);
+            UiExpToPercentage = LoadPreference("UiExpToPercentage", true);
+            UiHpToPercentage = LoadPreference("UiHpToPercentage", false);
+            UiMpToPercentage = LoadPreference("UiMpToPercentage", false);
```
Comment (lodicolo) [inline Intersect.Client.Framework/Database/GameDatabase.cs:86]: Use `nameof` instead of a string literal

File: Intersect.Client.Framework/Database/GameDatabase.cs [hunk 6]
```diff
@@ -95,6 +103,9 @@ public virtual void SavePreferences()
             SavePreference("NpcOverheadInfo", NpcOverheadInfo.ToString());
             SavePreference("PartyMemberOverheadInfo", PartyMemberOverheadInfo.ToString());
             SavePreference("PlayerOverheadInfo", PlayerOverheadInfo.ToString());
+            SavePreference("UiExpToPercentage", UiExpToPercentage.ToString());
+            SavePreference("UiHpToPercentage", UiHpToPercentage.ToString());
+            SavePreference("UiMpToPercentage", UiMpToPercentage.ToString());
```
Comment (lodicolo) [inline Intersect.Client.Framework/Database/GameDatabase.cs:108]: Use `nameof` instead of a string literal

File: Intersect.Client/Interface/Game/EntityPanel/EntityBox.cs [hunk 7]
```diff
@@ -585,10 +585,24 @@ private void UpdateHpBar(float elapsedTime, bool instant = false)
                 shieldfillRatio = Math.Min(1, Math.Max(0, shieldfillRatio));
                 targetShieldWidth = (float) Math.Floor(shieldfillRatio * width);
 
-                //Fix the Labels
-                HpLbl.Text = Strings.EntityBox.vital0val.ToString(
-                    MyEntity.Vital[(int) Vitals.Health], MyEntity.MaxVital[(int) Vitals.Health]
+                float hpPercentage = hpfillRatio * 100;
+                var hpPercentageText = hpPercentage == 0 ? $"{hpPercentage}%" : $"{hpPercentage:0.0}%";
```
Comment (lodicolo) [inline Intersect.Client/Interface/Game/EntityPanel/EntityBox.cs:589]: This ternary is unnecessary, use `{hpPercentage:0.##}%` instead.

File: Intersect.Client/Interface/Game/EntityPanel/EntityBox.cs [hunk 8]
```diff
@@ -684,14 +698,30 @@ private void UpdateHpBar(float elapsedTime, bool instant = false)
...
+            
+            if (MyEntity.MaxVital[(int)Vitals.Mana] > 0)
             {
-                targetMpWidth = MyEntity.Vital[(int) Vitals.Mana] / (float) MyEntity.MaxVital[(int) Vitals.Mana];
+                targetMpWidth = MyEntity.Vital[(int)Vitals.Mana] / (float)MyEntity.MaxVital[(int)Vitals.Mana];
                 targetMpWidth = Math.Min(1, Math.Max(0, targetMpWidth));
-                MpLbl.Text = Strings.EntityBox.vital1val.ToString(
-                    MyEntity.Vital[(int) Vitals.Mana], MyEntity.MaxVital[(int) Vitals.Mana]
+                float mpPercentage = targetMpWidth * 100;
+                var mpPercentageText = mpPercentage == 0 ? $"{mpPercentage}%" : $"{mpPercentage:0.0}%";
```
Comment (lodicolo) [inline Intersect.Client/Interface/Game/EntityPanel/EntityBox.cs:707]: This ternary is unnecessary, use `{mpPercentage:0.##}%` instead.

File: Intersect.Client/Interface/Game/EntityPanel/EntityBox.cs [hunk 9]
```diff
@@ -742,14 +772,29 @@ private void UpdateMpBar(float elapsedTime, bool instant = false)
...
-                                 (float) ((Player) MyEntity).GetNextLevelExperience();
 
-                ExpLbl.Text = Strings.EntityBox.expval.ToString(
-                    ((Player) MyEntity)?.Experience, ((Player) MyEntity)?.GetNextLevelExperience()
+            if (((Player)MyEntity).GetNextLevelExperience() > 0)
+            {
+                targetExpWidth = (float)((Player)MyEntity).Experience /
+                                 (float)((Player)MyEntity).GetNextLevelExperience();
+                float expPercentage = targetExpWidth * 100;
+                var expPercentageText = expPercentage == 0 ? $"{expPercentage}%" : $"{expPercentage:0.00}%";
```
Comment (lodicolo) [inline Intersect.Client/Interface/Game/EntityPanel/EntityBox.cs:781]: This ternary is unnecessary, use `{expPercentage:0.##}%` instead.

File: Intersect.Client/Interface/Shared/SettingsWindow.cs [hunk 10]
```diff
@@ -193,17 +199,25 @@ public SettingsWindow(Base parent, MainMenu mainMenu, EscapeMenu escapeMenu)
...
             mAutoCloseWindowsCheckbox.Text = Strings.Settings.AutoCloseWindows;
 
+            // Game Settings - Toggle for: User Interface EXP/HP/MP to Percentage.
+            mUiExpToPercentageCheckbox = new LabeledCheckBox(mGameSettingsContainer, "UiExpToPercentageCheckbox");
+            mUiExpToPercentageCheckbox.Text = Strings.Settings.UiExpToPercentage;
+            mUiHpToPercentageCheckbox = new LabeledCheckBox(mGameSettingsContainer, "UiHpToPercentageCheckbox");
+            mUiHpToPercentageCheckbox.Text = Strings.Settings.UiHpToPercentage;
+            mUiMpToPercentageCheckbox = new LabeledCheckBox(mGameSettingsContainer, "UiMpToPercentageCheckbox");
+            mUiMpToPercentageCheckbox.Text = Strings.Settings.UiMpToPercentage;
+
```
Comment (lodicolo) [inline Intersect.Client/Interface/Shared/SettingsWindow.cs:220]: The `Ui` prefix is not beneficial here at all, it's just two extra characters.

File: Intersect.Client/Localization/Strings.cs [hunk 11]
```diff
@@ -1787,6 +1787,15 @@ public partial struct Settings
 
             [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
             public static LocalizedString Vsync = @"V-Sync";
+            
+            [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
+            public static LocalizedString UiExpToPercentage = @"EXP to Percentage";
```
Comment (lodicolo) [inline Intersect.Client/Localization/Strings.cs:1792]: `Show Experience as percentage`

File: Intersect.Client/Localization/Strings.cs [hunk 12]
```diff
@@ -1787,6 +1787,15 @@ public partial struct Settings
...
+            
+            [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
+            public static LocalizedString UiHpToPercentage = @"HP to Percentage";
```
Comment (lodicolo) [inline Intersect.Client/Localization/Strings.cs:1795]: `Show Health as percentage`

File: Intersect.Client/Localization/Strings.cs [hunk 13]
```diff
@@ -1787,6 +1787,15 @@ public partial struct Settings
...
+            
+            [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
+            public static LocalizedString UiMpToPercentage = @"MP to Percentage";
```
Comment (lodicolo) [inline Intersect.Client/Localization/Strings.cs:1798]: `Show Mana as percentage`

File: Intersect.Client/Interface/Game/EntityPanel/EntityBox.cs [hunk 14]
```diff
@@ -742,14 +772,29 @@ private void UpdateMpBar(float elapsedTime, bool instant = false)
...
+                {
+                    ExpLbl.Text = expPercentageText;
+                    ExpBackground.SetToolTipText(expValueText);
+                    ExpBar.SetToolTipText(expValueText);
+                }
+                else
+                {
+                    ExpLbl.Text = expValueText;
+                    ExpBackground.SetToolTipText(expPercentageText);
+                    ExpBar.SetToolTipText(expPercentageText);
```
Comment (lodicolo) [inline Intersect.Client/Interface/Game/EntityPanel/EntityBox.cs:796]: Is setting both tooltips really necessary? Is there a difference in behavior if only one of them is set?

**Review by lodicolo — CHANGES_REQUESTED** [review]

File: Intersect.Client.Framework/Database/GameDatabase.cs (hunk 1 above)
Comment (Arufonsu) [reply Intersect.Client.Framework/Database/GameDatabase.cs:38]: My bad, at some point i meant to re-organize these variables according to their categories with comments, let me know if i should undo this one. as for the other properties, I've just added their missing `{ get; set; }`

File: Intersect.Client/Localization/Strings.cs (hunk 11 above)
Comment (Arufonsu) [reply Intersect.Client/Localization/Strings.cs:1792]: `Show Experience as percentage` it too long and hard to fit in the graphic image-box that belongs to the interface settings ; would you agree with `Experience as percentage` instead?

File: Intersect.Client.Framework/Database/GameDatabase.cs (hunk 6 above)
Comment (Arufonsu) [reply Intersect.Client.Framework/Database/GameDatabase.cs:108]: like this? :

image

should i change the rest within `SavePreferences()` as well?

File: Intersect.Client/Interface/Game/EntityPanel/EntityBox.cs (hunk 14 above)
Comment (Arufonsu) [reply Intersect.Client/Interface/Game/EntityPanel/EntityBox.cs:796]: I was doing this because bars and their backgrounds had `"MouseInputEnabled": true` , so if i added the tooltip to the background only, the bar wouldn't allow it to popup when being over the background.
today i've just realized that having the tooltip enabled for the background is more than enough, and in order for this to work i'll have to change the assets so the actual bars over the background have `MouseInputEnabled` disabled.
I'll change this in the next commit.

File: Intersect.Client/Localization/Strings.cs (hunk 12 above)
Comment (Arufonsu) [reply Intersect.Client/Localization/Strings.cs:1795]: https://github.com/AscensionGameDev/Intersect-Engine/pull/1448/files#r967848965

File: Intersect.Client/Localization/Strings.cs (hunk 13 above)
Comment (Arufonsu) [reply Intersect.Client/Localization/Strings.cs:1798]: https://github.com/AscensionGameDev/Intersect-Engine/pull/1448/files#r967848965

File: Intersect.Client.Framework/Database/GameDatabase.cs (hunk 1 above)
Comment (Arufonsu) [reply Intersect.Client.Framework/Database/GameDatabase.cs:38]: moved back `HideOthersOnWindowOpen`

**Comment by Arufonsu** [comment]
Updated with review, [also updated the assets](https://github.com/AscensionGameDev/Intersect-Assets/pull/22) due to some changes.

Preview:

https://user-images.githubusercontent.com/17498701/189541998-d93331f3-e59d-4c0f-bfb6-f52688689db5.mp4

**Comment by Arufonsu** [comment]
Updated with panda's review, [also updated the assets](https://github.com/AscensionGameDev/Intersect-Assets/pull/22) due to some changes.

Preview:

image

**Review by lodicolo — APPROVED** [review]
